### PR TITLE
feat(targeting): add targetingContextProvider

### PR DESCRIPTION
**Pull request by azuruce**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @azuruce has signed the CLA. Thanks for the pull request!

**Comment by azuruce** [comment]
looking for early review

**Comment by mickr** [comment]
I would update the title of this PR to match the first commit just so there isn't confusion if you squash and merge.

**Comment by azuruce** [comment]
@mickr Thanks for the review! I will merge first two commits separately in #1844 . This PR should only merge the last commit and it is currently missing test but I posted to get early review as I am ramping up on box-ui-elements contribution.

File: src/features/targeting/types.js [hunk 1]
```diff
@@ -1,35 +1,8 @@
...
-import type { SetEligibleMessageHeaders, TargetingApi } from './targeting-api';
 import { MESSAGE_TYPES } from './constants';
 
 export type MessageType = $Values<typeof MESSAGE_TYPES>;
 
 export type MessageID = number;
 
 export type MessageName = string;
-
-export type MessageHeader = {|
```
Comment (azuruce) [inline src/features/targeting/types.js]: I removed those fields as I figured out I don't need them in targeting context. Those fields stay with in app messenger.

File: src/features/targeting/README.md [hunk 2]
```diff
@@ -9,12 +9,13 @@ Targeting Api is provided for each message. It can be consumed by the below pseu
 function SharedLinkPreviewModal() {
   const { getTargetingApi } = useContext(TargetingContext);
 
-  const { shouldShow, onClose, onSeen } = getTargetingApi('adhoc_shared_link_preview');
+  const { shouldShow, onClosed, onSeen } = getTargetingApi('adhoc_shared_link_preview');
```
Comment (laughnan) [inline src/features/targeting/README.md]: i tend to prefer `toClose` as it's the verb and makes sense in this context. realize it's a little weird with `onSeen` (which is also a verb, but in the past tense) but we're doing `toClose` in the present tense.

Comment (laughnan) [reply src/features/targeting/README.md]: i'm sorry, i meant to say `onClose` (not `toClose`). that was a typo on my part.

File: src/features/targeting/contexts.js [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
+|};
 
 const FtuxContext = createContext<FtuxContextValue>({});
 
 const defaultTargetingContextValue = {
     getTargetingApi: () => ({
         shouldShow: () => false,
         onSeen: () => {},
-        onClosed: () => {},
+        toClose: () => {},
```
Comment (laughnan) [inline src/features/targeting/contexts.js]: sorry can we make this `onClose` (i'm so sorry that my comment previously was inaccurate).

File: src/features/targeting/targeting-api.js [hunk 4]
```diff
@@ -1,23 +1,15 @@
...
-    onClosed: () => void,
-    // call to onSeen will cause the message seen count to increment
+    // call onSeen after message is shown
     onSeen: () => void,
-    // call to onClosed will cause shouldShow to be false for the rest of the session, and
-    // message close count to increment.
+    // return true if message is eligible to show and not yet closed
     shouldShow: () => boolean,
+    // call toClose just before message should be closed, to cause shouldShow to return false
+    toClose: () => void,
```
Comment (laughnan) [inline src/features/targeting/targeting-api.js]: sorry can we make this `onClose` (i'm so sorry that my comment previously was inaccurate).

File: src/components/targeting/__tests__/TargetingContextProvider.test.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+// @flow
+import * as React from 'react';
+import { act } from 'react-dom/test-utils';
+
+import TargetingContextProvider from '../TargetingContextProvider';
+import { TargetingContext } from '../../../features/targeting';
+
+function Consumer() {
+    const { getTargetingApi } = React.useContext(TargetingContext);
+    const { shouldShow, toClose, onSeen } = getTargetingApi('msg');
```
Comment (laughnan) [inline src/components/targeting/__tests__/TargetingContextProvider.test.js]: sorry can we make this `onClose` (i'm so sorry that my comment previously was inaccurate).

File: src/features/targeting/README.md (hunk 2 above)
Comment (azuruce) [reply src/features/targeting/README.md]: hi, I think toClose is better in terms of it is clear this should be called hand. 
shouldShow() // boolean to show
onSeen() // call after seen
toClose() // call before close

Comment (azuruce) [reply src/features/targeting/README.md]: fixed per review

File: src/features/targeting/useMessage.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    return {
+        canShow: name in eligibleMessageIDMap && messageStateMap[name] !== CLOSED,
+        onShow: () => {
+            if (!messageStateMap[name]) {
+                setMessageStateMap({ ...messageStateMap, [name]: SHOWING });
+                messageApi.markMessageAsSeen(eligibleMessageIDMap[name]);
+            }
+        },
+        onClose: () => {
+            if (messageStateMap[name] === 'showing') {
```
Comment (DanDeMicco) [inline src/features/targeting/useMessage.js]: use constant here

File: src/features/targeting/useMessage.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import type { TargetingApi } from './TargetingApi';
+import MessageContext from './MessageContext';
+
+const useMessage = (name: MessageName): TargetingApi => {
+    const { eligibleMessageIDMap, messageApi, messageStateMap, setMessageStateMap } = useContext(MessageContext);
+
+    return {
+        canShow: name in eligibleMessageIDMap && messageStateMap[name] !== CLOSED,
+        onShow: () => {
+            if (!messageStateMap[name]) {
```
Comment (DanDeMicco) [inline src/features/targeting/useMessage.js]: should you check if canShow?

**Comment by DanDeMicco** [comment]
looking good just minor feedback

File: src/features/targeting/TargetingApi.js [hunk 8]
```diff
@@ -4,11 +4,15 @@
  * basis. Currently, useMessage can provide targeting from MessageContext.
  */
 
+import type { MessageEvent } from './types';
+
 export type TargetingApi = $ReadOnly<{
     // is true if message is eligible to show and not yet closed
-    canShow: boolean,
-    // call onClose just before message should be closed, to cause canShow to be false
+    canShow: () => boolean,
```
Comment (azuruce) [inline src/features/targeting/TargetingApi.js]: convert canShow back to function as in a State Machine situation such as ftux, the canShow comes from a function injected by state machine

File: src/features/targeting/constants.js [hunk 9]
```diff
@@ -4,4 +4,8 @@ const CLOSED: 'closed' = 'closed';
 const SHOWING: 'showing' = 'showing';
 const MESSAGE_STATES = { CLOSED, SHOWING };
 
-export { CLOSED, SHOWING, MESSAGE_STATES };
+const SHOW: 'show' = 'show';
+const CLOSE: 'close' = 'close';
+const MESSAGE_EVENTS = { SHOW, CLOSE };
```
Comment (azuruce) [inline src/features/targeting/constants.js]: define events, so state machine can receive events as input.

**Pull request merged**